src/frame-validator.cpp

```diff
@@ -3,90 +3,106 @@
 
 #include "frame-validator.h"
 
+#define INVALID_PIXELS_THRESHOLD 0.1
+
 namespace librealsense
 {
     void frame_validator::on_frame(rs2_frame * f)
     {
-        if (propagate((frame_interface*)f))
+        if (!_stopped && propagate((frame_interface*)f))
         {
-            _user_callback->on_frame(f);
-            _sensor->set_frames_callback(_user_callback);
+            if(is_user_requseted_frame((frame_interface*)f))
+                _user_callback->on_frame(f);
         }
     }
 
     void frame_validator::release()
-    {
-
-    }
+    {}
 
-    frame_validator::frame_validator(std::shared_ptr<sensor_interface> sensor, frame_callback_ptr user_callback, stream_profiles current_requests) :
+    frame_validator::frame_validator(std::shared_ptr<sensor_interface> sensor, frame_callback_ptr user_callback, stream_profiles current_requests, stream_profiles validator_requests) :
         _sensor(sensor), 
         _user_callback(user_callback), 
-        _current_requests(current_requests)
-    {
-    
-    }
+        _user_requests(current_requests),
+        _validator_requests(validator_requests),
+        _stopped(false),
+        _validated(false)
+    {}
 
     frame_validator::~frame_validator()
-    {
-
-    }
+    {}
 
     bool frame_validator::propagate(frame_interface* frame)
     {
-        auto vf = dynamic_cast<video_frame*>(frame);
-        if (vf == nullptr) {
+        if(_validated)
             return true;
-        }
+
+        auto vf = dynamic_cast<video_frame*>(frame);
+        if (vf == nullptr)
+            throw std::runtime_error(to_string() << "non video stream arrived to frame_validator");
+
         auto stream = vf->get_stream();
-        if (stream->get_stream_type() != RS2_STREAM_DEPTH)
+
+        //all streams except ir will be droped untill the validation on ir,
+        //after the validation all streams will be passeded to user callback directly
+        if (stream->get_stream_type() != RS2_STREAM_INFRARED)
+            return false;
+
+        //start to validate only from the second frame
+        if(frame_num++ < 2)
             return false;
 
         auto w = vf->get_width();
         auto h = vf->get_height();
-        auto bpp = vf->get_bpp() / 8;
-        auto data = reinterpret_cast<const uint16_t*>(vf->get_frame_data());
 
-        bool non_empty_pixel_found = false;
+        auto data = static_cast<const void*>(vf->get_frame_data());
+
+        auto invalid_pixels = 0;
 
-        int y_begin = h * 0.5 - h * 0.05;
-        int y_end = h * 0.5 + h * 0.05;
-        int x_begin = w * 0.5 - w * 0.05;
-        int x_end = w * 0.5 + w * 0.05;
-        for (int y = y_begin; y < y_end; y++)
+        for (int y = 0; y < h*w; y++)
         {
-            for (int x = x_begin; x < x_end; x++)
+            if (((byte*)data)[y] == 0)
```

Comment:
Is IR stream 8-bit only, no 12 bit formats?

Reply:
yes 8 bit